wgpu-core/src/device/any_device.rs

```diff
@@ -1,72 +1,98 @@
+        if self.vtable.backend != A::VARIANT {
+            return None;
+        }
+
+        // SAFETY: We just checked the instance above implicitly by the backend
+        // that it was statically constructed through.
+        Some(unsafe { &*(self.data.as_ptr().cast::<Device<A>>()) })
     }
 
     /// If `self` is an `Arc<Device<A>>`, return a clone of that.
     pub fn downcast_clone<A: HalApi>(&self) -> Option<Arc<Device<A>>> {
-        // `Arc::downcast` returns `Arc<T>`, but requires that `T` be `Sync` and
-        // `Send`, and this is not the case for `Device` in wasm builds.
-        //
-        // But as far as I can see, `Arc::downcast` has no particular reason to
-        // require that `T` be `Sync` and `Send`; the steps used here are sound.
-        if (self.0).is::<Device<A>>() {
-            // Get an owned Arc.
-            let clone = self.0.clone();
-            // Turn the `Arc`, which is a pointer to an `ArcInner` struct, into
-            // a pointer to the `ArcInner`'s `data` field. Carry along the
-            // vtable from the original `Arc`.
-            let raw_erased: *const (dyn Any + 'static) = Arc::into_raw(clone);
-            // Remove the vtable, and supply the concrete type of the `data`.
-            let raw_typed: *const Device<A> = raw_erased.cast::<Device<A>>();
-            // Convert the pointer to the `data` field back into a pointer to
-            // the `ArcInner`, and restore reference-counting behavior.
-            let arc_typed: Arc<Device<A>> = unsafe {
-                // Safety:
-                // - We checked that the `dyn Any` was indeed a `Device<A>` above.
-                // - We're calling `Arc::from_raw` on the same pointer returned
-                //   by `Arc::into_raw`, except that we stripped off the vtable
-                //   pointer.
-                // - The pointer must still be live, because we've borrowed `self`,
-                //   which holds another reference to it.
-                // - The format of a `ArcInner<dyn Any>` must be the same as
-                //   that of an `ArcInner<Device<A>>`, or else `AnyDevice::new`
-                //   wouldn't be possible.
-                Arc::from_raw(raw_typed)
-            };
-            Some(arc_typed)
-        } else {
-            None
+        if self.vtable.backend != A::VARIANT {
+            return None;
         }
+
+        // We need to prevent the destructor of the arc from running, since it
+        // refers to the instance held by this object. Dropping it would
+        // invalidate this object.
+        //
+        // SAFETY: We just checked the instance above implicitly by the backend
+        // that it was statically constructed through.
+        let this =
+            ManuallyDrop::new(unsafe { Arc::from_raw(self.data.as_ptr().cast::<Device<A>>()) });
+
+        // Cloning it increases the reference count, and we return a new arc
+        // instance.
+        Some((*this).clone())
```

Comment:
nitpick: I think it would be clearer to use [`Arc::increment_strong_count`](https://doc.rust-lang.org/stable/std/sync/struct.Arc.html#method.increment_strong_count), despite technically being yet another `unsafe` call. Ignoring my feedback about changing `NonNull<()>` to be a `*mut c_void`, this would look like:

```suggestion        let ptr = self.data.as_ptr().cast::<Device<A>>();
        // Increment the strong count on this pointer so we can give a "cloned" `Arc` back.
        //
        // SAFETY:
        //
        // * We know that `ptr` points to a `Device<A>` because of the above check.
        // * We also know the strong reference count should be greater than 0, because, in this
        //   API, we've only allowed either (a) references to the pointee, or (b) `Arc`s with their
        //   own corresponding strong reference count to decrement on `Drop`.
        unsafe { Arc::increment_strong_count(ptr) };
        // SAFETY: We already assume we have a `Device<A>` above, so this isn't adding unsoundness.
        Some(unsafe { Arc::from_raw(ptr) })

```